pytket/extensions/cuquantum/mps/mps.py

```diff
@@ -0,0 +1,639 @@
+            # Initialise the tensor to ket 0
+            m_tensor[0][0][0] = 1
+            m_tensor[0][0][1] = 0
+            self.tensors.append(Tensor(m_tensor, [i, i + 1, i + n_tensors]))
+
+        # Append the rightmost tensor
+        self.tensors.append(Tensor(r_tensor, [n_tensors - 1, 2 * n_tensors - 1]))
+
+    def is_valid(self) -> bool:
+        """Verify that the MPS does not exceed the dimension limit (chi) of
+        the virtual bonds, that physical bonds have dimension 2 and that
+        the virtual bonds are connected in a line.
+
+        Returns:
+            False if a violation was detected.
+        """
+        self._flush()
+
+        chi_ok = all(
+            all(dim <= self.chi for dim in self.get_virtual_dimensions(pos))
+            for pos in range(len(self))
+        )
+        phys_ok = all(self.get_physical_dimension(pos) == 2 for pos in range(len(self)))
+        shape_ok = all(
+            len(tensor.data.shape) == len(tensor.bonds) and len(tensor.bonds) <= 3
+            for tensor in self.tensors
+        )
+
+        v_bonds_ok = True
+        # Check the leftmost tensor
+        v_bonds_ok = v_bonds_ok and self.get_virtual_bonds(0)[0] == 1
+        # Check the middle tensors
+        for i in range(1, len(self) - 1):
+            v_bonds_ok = v_bonds_ok and (
+                self.get_virtual_bonds(i)[0] == i
+                and self.get_virtual_bonds(i)[1] == i + 1
+            )
+        # Check the rightmost tensor
+        i = len(self) - 1
+        v_bonds_ok = v_bonds_ok and self.get_virtual_bonds(i)[0] == i
+
+        return chi_ok and phys_ok and shape_ok and v_bonds_ok
+
+    def apply_gate(self, gate: Command) -> MPS:
+        """Apply the gate to the MPS.
+
+        Note:
+            Only one-qubit gates and two-qubit gates are supported. Two-qubit
+            gates must act on adjacent qubits.
+
+        Args:
+            gate: The gate to be applied.
+
+        Returns:
+            ``self``, to allow for method chaining.
+        """
+        positions = [self.qubit_position[q] for q in gate.qubits]
+
+        if len(positions) == 1:
+            self._apply_1q_gate(positions[0], gate.op)
+
+        elif len(positions) == 2:
+            dist = positions[1] - positions[0]
+            # We explicitly allow both dist==1 or dist==-1 so that non-symmetric
+            # gates such as CX can use the same Op for the two ways it can be in.
+            if dist not in [1, -1]:
+                raise RuntimeError(
+                    "Gates must be applied to adjacent qubits! "
+                    + f"This is not satisfied by {gate}."
+                )
+            self._apply_2q_gate(positions, gate.op)
+
+        else:
+            raise RuntimeError(
+                "Gates must act on only 1 or 2 qubits! "
+                + f"This is not satisfied by {gate}."
+            )
+        return self
+
+    def vdot(self, mps: MPS) -> complex:

```

Comment:
I have noticed that I am using a different convention here than the current `vdot` in `TensorNetwork` from our `develop` branch: I'm obtaining the conjugate of `self`, but in `develop` we take the conjugate of `other_tn`. As a consequence, what I return would be the conjugate of the complex number we'd get if we adhere to the convention in `develop`.

However, notice that https://numpy.org/doc/stable/reference/generated/numpy.vdot.html takes the conjugate of the _first_ argument. Should we consider `self` as the first argument and, hence, use the convention I use here?

Either way, I should clarify in the docstring which of the two is conjugated.

Reply:
Docstring clarified. The question of which state to conjugate remains open.

Reply:
I agree, we should modify `vdot` in the converter as per numpy convention. No idea why I did it another way around (ah, probably because I was considering self as a ket, and other as a bra). But I also don't think it is urgent - the two classes don't speak to each other.